File: tests/test_serializable_collection.py
# SPDX-License-Identifier: Apache-2.0
# Copyright Contributors to the OpenTimelineIO project

import unittest

import opentimelineio as otio
import opentimelineio.test_utils as otio_test_utils


class SerializableColTests(unittest.TestCase, otio_test_utils.OTIOAssertions):
    def setUp(self):
        self.maxDiff = None
        self.children = [
            otio.schema.Clip(name="testClip"),
            otio.schema.MissingReference()
        ]
        self.md = {'foo': 'bar'}
        self.sc = otio.schema.SerializableCollection(
            name="test",
            children=self.children,
            metadata=self.md
        )

    def test_ctor(self):
        self.assertEqual(self.sc.name, "test")
        self.assertEqual(self.sc[:], self.children)
        self.assertEqual(self.sc.metadata, self.md)

    def test_iterable(self):
        self.assertEqual(self.sc[0], self.children[0])
        self.assertEqual([i for i in self.sc], self.children)
        self.assertEqual(len(self.sc), 2)

        # test recursive iteration
        sc = otio.schema.SerializableCollection(
            name="parent",
            children=[self.sc]
        )

        self.assertEqual(len(list(sc.each_clip())), 1)

        # test deleting an item
        tmp = self.sc[0]
        del self.sc[0]
        self.assertEqual(len(self.sc), 1)
        self.sc[0] = tmp
        self.assertEqual(self.sc[0], tmp)

    def test_serialize(self):
        encoded = otio.adapters.otio_json.write_to_string(self.sc)
        decoded = otio.adapters.otio_json.read_from_string(encoded)
        self.assertIsOTIOEquivalentTo(self.sc, decoded)

    def test_str(self):
        self.assertMultiLineEqual(
            str(self.sc),
            "SerializableCollection(" +
            str(self.sc.name) + ", " +
            str(list(self.sc)) + ", " +
            str(self.sc.metadata) +
            ")"
        )

    def test_repr(self):
        self.assertMultiLineEqual(
            repr(self.sc),
            "otio.schema.SerializableCollection(" +
            "name=" + repr(self.sc.name) + ", " +
            "children=" + repr(list(self.sc)) + ", " +
            "metadata=" + repr(self.sc.metadata) +
            ")"
        )

    def test_children_if(self):
        cl = otio.schema.Clip()
        tr = otio.schema.Track()
        tr.append(cl)
        tl = otio.schema.Timeline()
        tl.tracks.append(tr)
        sc = otio.schema.SerializableCollection()
        sc.append(tl)
        result = sc.children_if(otio.schema.Clip)
        self.assertEqual(len(result), 1)
        self.assertEqual(result[0], cl)

    def test_children_if_search_range(self):
        range = otio.opentime.TimeRange(
            otio.opentime.RationalTime(0.0, 24.0),
            otio.opentime.RationalTime(24.0, 24.0))
        cl0 = otio.schema.Clip()
        cl0.source_range = range
        cl1 = otio.schema.Clip()
        cl1.source_range = range
        cl2 = otio.schema.Clip()
        cl2.source_range = range
        tr = otio.schema.Track()
        tr.append(cl0)
        tr.append(cl1)
        tr.append(cl2)
        tl = otio.schema.Timeline()
        tl.tracks.append(tr)
        sc = otio.schema.SerializableCollection()
        sc.append(tl)
        result = sc.children_if(otio.schema.Clip, range)
        self.assertEqual(len(result), 1)
        self.assertEqual(result[0], cl0)

    def test_children_if_shallow_search(self):
        cl = otio.schema.Clip()
        tr = otio.schema.Track()
        tr.append(cl)
        tl = otio.schema.Timeline()
        tl.tracks.append(tr)
        sc = otio.schema.SerializableCollection()
        sc.append(tl)
        result = sc.children_if(otio.schema.Clip, shallow_search=True)
        self.assertEqual(len(result), 0)
        result = sc.children_if(otio.schema.Clip, shallow_search=False)
        self.assertEqual(len(result), 1)
        self.assertEqual(result[0], cl)


if __name__ == '__main__':
    unittest.main()

File: tests/test_serializableCollection.cpp
// SPDX-License-Identifier: Apache-2.0
// Copyright Contributors to the OpenTimelineIO project

#include "utils.h"

#include <opentimelineio/clip.h>
#include <opentimelineio/serializableCollection.h>
#include <opentimelineio/timeline.h>
#include <opentimelineio/track.h>

#include <iostream>

using namespace OTIO_NS;

int
main(int argc, char** argv)
{
    Tests tests;

    tests.add_test(
        "test_find_children", [] {
        SerializableObject::Retainer<Clip> cl = new Clip();
        SerializableObject::Retainer<Track> tr = new Track();
        tr->append_child(cl);
        SerializableObject::Retainer<Timeline> tl = new Timeline();
        tl->tracks()->append_child(tr);
        SerializableObject::Retainer<SerializableCollection>
            sc = new SerializableCollection();
        sc->insert_child(0, tl);
        OTIO_NS::ErrorStatus err;
        auto result = sc->find_children<Clip>(&err, {}, false);
        assertEqual(result.size(), 1);
        assertEqual(result[0].value, cl.value);
    });
    tests.add_test(
        "test_find_children_search_range", [] {
        const TimeRange range(RationalTime(0.0, 24.0), RationalTime(24.0, 24.0));
        SerializableObject::Retainer<Clip> cl0 = new Clip();
        cl0->set_source_range(range);
        SerializableObject::Retainer<Clip> cl1 = new Clip();
        cl1->set_source_range(range);
        SerializableObject::Retainer<Clip> cl2 = new Clip();
        cl2->set_source_range(range);
        SerializableObject::Retainer<Track> tr = new Track();
        tr->append_child(cl0);
        tr->append_child(cl1);
        tr->append_child(cl2);
        SerializableObject::Retainer<Timeline> tl = new Timeline();
        tl->tracks()->append_child(tr);
        SerializableObject::Retainer<SerializableCollection>
            sc = new SerializableCollection();
        sc->insert_child(0, tl);
        OTIO_NS::ErrorStatus err;
        auto result = sc->find_children<Clip>(&err, range);
        assertEqual(result.size(), 1);
        assertEqual(result[0].value, cl0.value);
    });
    tests.add_test(
        "test_find_children_shallow_search", [] {
        SerializableObject::Retainer<Clip> cl = new Clip();
        SerializableObject::Retainer<Track> tr = new Track();
        tr->append_child(cl);
        SerializableObject::Retainer<Timeline> tl = new Timeline();
        tl->tracks()->append_child(tr);
        SerializableObject::Retainer<SerializableCollection>
            sc = new SerializableCollection();
        sc->insert_child(0, tl);
        OTIO_NS::ErrorStatus err;
        auto result = sc->find_children<Clip>(&err, std::nullopt, true);
        assertEqual(result.size(), 0);
        result = sc->find_children<Clip>(&err, std::nullopt, false);
        assertEqual(result.size(), 1);
        assertEqual(result[0].value, cl.value);
    });

    tests.run(argc, argv);
    return 0;
}
